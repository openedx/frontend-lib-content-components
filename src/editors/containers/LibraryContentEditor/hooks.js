import {
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useDispatch } from 'react-redux';

import { modes } from './constants';
import { actions } from '../../data/redux';
import * as requests from './data/requests';
import { RequestKeys, RequestStates } from '../../data/constants/requests';
import { isV1Library, getCandidates, getSelectedRows } from './utils';

export const blockTypeDisplay = (type) => {
  if (type === 'html') { return 'Text'; }
  if (type === 'video') { return 'Video'; }
  if (type === 'problem') { return 'Problem'; }
  return 'Other';
};

export const useLibraryHook = ({
  blockFailed,
  blockFinished,
  blockValue,
}) => {
  const dispatch = useDispatch();

  // load previously saved block data from blockValue
  useEffect(() => {
    const metadata = blockValue?.data?.metadata;
    const libraryId = metadata?.source_library_id ?? null;
    let settings = {};
    if (libraryId) {
      settings = {
        [libraryId]: {
          version: metadata?.source_library_version,
          mode: metadata?.manual ? modes.selected.value : modes.random.value,
          count: metadata?.max_count,
          showReset: metadata?.allow_resetting_children,
          candidates: metadata?.candidates,
          blocks: [],
        },
      };
    }
    dispatch(actions.library.initializeFromBlockValue({ libraryId, settings }));
  }, [blockValue]);

  // fetch libraries when block finishes loading
  useEffect(() => {
    if (blockFinished && !blockFailed) {
      dispatch(requests.fetchV2Libraries({
        onSuccess: (response) => {
          const libraries = {};
          response?.data.forEach(library => {
            libraries[library.id] = library;
          });
          dispatch(actions.library.addLibraries({ libraries }));
        },
        onFailure: (error) => {
          dispatch(actions.requests.failRequest({
            requestKey: RequestKeys.fetchV2Libraries,
            error,
          }));
        },
      }));
      dispatch(requests.fetchV1Libraries({
        onSuccess: (response) => {
          const libraries = {};
          response?.data?.libraries.forEach(library => {
            libraries[library.library_key] = library;
          });
          dispatch(actions.library.addLibraries({ libraries }));
        },
        onFailure: (error) => {
          dispatch(actions.requests.failRequest({
            requestKey: RequestKeys.fetchV1Libraries,
            error,
          }));
        },
      }));
    }
  }, [blockFinished, blockFailed]);
};

export const useLibrarySelectorHook = ({
  selectedLibraryId,
  settings,
}) => {
  const dispatch = useDispatch();

  return {
    onLibrarySelect: (id) => {
      if (id !== selectedLibraryId) {
        dispatch(actions.library.setLibraryId({ selectedLibraryId: id }));
        if (!settings[id]) {
          dispatch(actions.library.initialLibrarySettings({ selectedLibraryId: id }));
        }
      }
    },
  };
};

export const useBlocksSelectorHook = ({
  blocks,
  candidates,
  libraries,
  savedLibraryId,
  selectedLibraryId,
  v1BlockRequests,
}) => {
  const dispatch = useDispatch();

  const [tableDataLoaded, setTableDataLoaded] = useState(false);

  // fetch library version and blocks
  useEffect(() => {
    setTableDataLoaded(false);
    if (!blocks || blocks.length === 0) {
      if (selectedLibraryId === savedLibraryId) {
        dispatch(requests.fetchChildrenInfo({
          onSuccess: (response) => {
            const children = response?.data?.children;
            dispatch(actions.library.setLibraryBlocks({
              blocks: children.map(block => ({
                id: block.id,
                display_name: block.display_name,
                block_type: block.category,
              })),
            }));
          },
          onFailure: (error) => {
            dispatch(actions.requests.failRequest({
              requestKey: RequestKeys.fetchChildrenInfo,
              error,
            }));
          },
        }));
      } else if (isV1Library(selectedLibraryId)) {
        if (!v1BlockRequests || Object.keys(v1BlockRequests).length === 0) {
          dispatch(requests.fetchV1LibraryContent({
            libraryId: selectedLibraryId,
            onSuccess: (response) => {
              const v1Blocks = response.data?.blocks ?? [];
              const initialRequestStatus = {};
              v1Blocks.forEach(id => {
                initialRequestStatus[id] = RequestStates.inactive;
              });
              dispatch(actions.library.setLibraryVersion({
                version: response.data?.version,
              }));
              dispatch(actions.library.setV1BlockRequests({
                v1BlockRequests: initialRequestStatus,
              }));
            },
            onFailure: (error) => {
              dispatch(actions.requests.failRequest({
                requestKey: RequestKeys.fetchV1LibraryContent,
                error,
              }));
            },
          }));
        }
      } else {
        dispatch(actions.library.setLibraryVersion({
          version: libraries[selectedLibraryId]?.version,
        }));
        dispatch(requests.fetchV2LibraryContent({
          libraryId: selectedLibraryId,
          onSuccess: (response) => {
            dispatch(actions.library.setLibraryBlocks({
              blocks: response?.data,
            }));
          },
          onFailure: (error) => {
            dispatch(actions.requests.failRequest({
              requestKey: RequestKeys.fetchV2LibraryContent,
              error,
            }));
          },
        }));
      }
    }
  }, [selectedLibraryId]);

  // set up blocks for v1 library
  useEffect(() => {
    if (blocks.length < Object.keys(v1BlockRequests).length) {
      if (isV1Library(selectedLibraryId)) {
        Object.keys(v1BlockRequests).forEach(blockId => {
          const status = v1BlockRequests[blockId];
          if (status === RequestStates.inactive) {
            dispatch(actions.library.updateV1BlockRequestStatus({
              blockId,
              status: RequestStates.pending,
            }));
            dispatch(requests.fetchV1LibraryBlock({
              blockId,
              onSuccess: (response) => {
                const v1Block = {
                  id: blockId,
                  display_name: response.data?.display_name ?? '',
                  block_type: response.data?.category ?? '',
                };
                dispatch(actions.library.addLibraryBlock({ block: v1Block }));
                dispatch(actions.library.updateV1BlockRequestStatus({
                  blockId,
                  status: RequestStates.completed,
                }));
              },
              onFailure: () => {
                dispatch(actions.library.updateV1BlockRequestStatus({
                  blockId,
                  status: RequestStates.failed,
                }));
              },
            }));
          }
        });
      }
    }
  }, [v1BlockRequests]);

  // check if blocks for table is loaded
  useEffect(() => {
    let loaded = true;
    Object.keys(v1BlockRequests).forEach(blockId => {
      const status = v1BlockRequests[blockId];
      if ((status === RequestStates.inactive) || (status === RequestStates.pending)) {
        loaded = false;
      }
    });
    setTableDataLoaded(loaded);
  }, [v1BlockRequests]);

  return {
    tableDataLoaded,
    data: useMemo(() => {
      if (tableDataLoaded) {
        return blocks.map(block => ({
          display_name: block.display_name,
          block_type: blockTypeDisplay(block.block_type),
        }));
      }
      return [];
    }, [blocks, tableDataLoaded]),

    initialRows: useMemo(() => {
      if (tableDataLoaded) {
        return getSelectedRows({ blocks, candidates });
      }
      return {};
    }, [blocks, candidates, tableDataLoaded]),

    onSelectedRowsChanged: useCallback((selected) => {
      if (tableDataLoaded) {
        dispatch(actions.library.setCandidatesForLibrary({
          candidates: getCandidates({
            blocks,
            rows: selected,
          }),
        }));
      }
    }, [blocks, tableDataLoaded]),
  };
};
