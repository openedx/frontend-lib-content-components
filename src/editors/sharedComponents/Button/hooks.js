/* eslint-disable import/prefer-default-export */
export const getButtonProps = ({ variant, className }) => {
  const variantClasses = {
    default: '',
    add: 'pl-0 text-primary-500 button-variant-add',
  };
  const variantMap = {
    add: 'tertiary',
  };
  const classes = [variantClasses[variant]];
  if (className) { classes.push(className); }

  return {
    className: classes.join(' '),
    variant: variantMap[variant] || variant,
  };
};
