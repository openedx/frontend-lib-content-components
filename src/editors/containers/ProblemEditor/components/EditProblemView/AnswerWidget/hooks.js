import { useState, useEffect } from 'react';
import { StrictDict } from '../../../../../utils';
import * as module from './hooks';
import { actions } from '../../../../../data/redux';
import { ProblemTypeKeys } from '../../../../../data/constants/problem';

export const state = StrictDict({
  isFeedbackVisible: (val) => useState(val),
});

export const removeAnswer = ({ answer, dispatch }) => () => {
  dispatch(actions.problem.deleteAnswer({ id: answer.id }));
};

export const setAnswer = ({ answer, hasSingleAnswer, dispatch }) => (payload) => {
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, ...payload }));
};

export const prepareFeedback = (answer) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = module.state.isFeedbackVisible(false);
  useEffect(() => {
    // Show feedback fields if feedback is present
    const isVisible = !!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback;
    setIsFeedbackVisible(isVisible);
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    if (!!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };
  return {
    isFeedbackVisible,
    toggleFeedback,
  };
};

export const isSingleAnswerProblem = (problemType) => (
  problemType === ProblemTypeKeys.DROPDOWN || problemType === ProblemTypeKeys.SINGLESELECT
);

export const useAnswerContainer = ({ answers, updateField }) => {
  useEffect(() => {
    let answerCount = 0;
    answers.forEach(answer => {
      if (answer.correct) {
        answerCount += 1;
      }
    });
    updateField({ correctAnswerCount: answerCount });
  }, []);
};

export default {
  state, removeAnswer, setAnswer, prepareFeedback, isSingleAnswerProblem, useAnswerContainer,
};
