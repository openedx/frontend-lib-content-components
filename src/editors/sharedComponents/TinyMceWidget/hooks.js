import {
  useState,
  useRef,
  useCallback,
  useEffect,
} from 'react';
import { getLocale, isRtl } from '@edx/frontend-platform/i18n';
import { a11ycheckerCss } from 'frontend-components-tinymce-advanced-plugins';
import tinyMCEStyles from '../../data/constants/tinyMCEStyles';
import { StrictDict } from '../../utils';
import pluginConfig from './pluginConfig';
import * as module from './hooks';
import tinyMCE from '../../data/constants/tinyMCE';

export const state = StrictDict({
  // eslint-disable-next-line react-hooks/rules-of-hooks
  isImageModalOpen: (val) => useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  isSourceCodeModalOpen: (val) => useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  imageSelection: (val) => useState(val),
  // eslint-disable-next-line react-hooks/rules-of-hooks
  refReady: (val) => useState(val),
});

export const addImagesAndDimensionsToRef = ({ imagesRef, assets, editorContentHtml }) => {
  const imagesWithDimensions = module.filterAssets({ assets }).map((image) => {
    const imageFragment = module.getImageFromHtmlString(editorContentHtml, image.url);
    return { ...image, width: imageFragment?.width, height: imageFragment?.height };
  });

  imagesRef.current = imagesWithDimensions;
};

export const useImages = ({ assets, editorContentHtml }) => {
  const imagesRef = useRef([]);

  useEffect(() => {
    module.addImagesAndDimensionsToRef({ imagesRef, assets, editorContentHtml });
  }, []);

  return { imagesRef };
};

export const parseContentForLabels = ({ editor, updateContent }) => {
  let content = editor.getContent();
  if (content && content?.length > 0) {
    const parsedLabels = content.split(/<label>|<\/label>/gm);
    let updatedContent;
    parsedLabels.forEach((label, i) => {
      if (!label.startsWith('<') && !label.endsWith('>')) {
        let previousLabel = parsedLabels[i - 1];
        let nextLabel = parsedLabels[i + 1];
        if (!previousLabel.endsWith('<p>')) {
          previousLabel = `${previousLabel}</p><p>`;
          updatedContent = content.replace(parsedLabels[i - 1], previousLabel);
          content = updatedContent;
          updateContent(content);
        }
        if (!nextLabel.startsWith('</p>')) {
          nextLabel = `</p><p>${nextLabel}`;
          updatedContent = content.replace(parsedLabels[i + 1], nextLabel);
          content = updatedContent;
          updateContent(content);
        }
      }
    });
  } else {
    updateContent(content);
  }
};

export const replaceStaticwithAsset = ({
  editor,
  imageUrls,
  editorType,
  lmsEndpointUrl,
  updateContent,
}) => {
  let content = editor.getContent();
  const imageSrcs = content.split('src="');
  imageSrcs.forEach(src => {
    const currentContent = content;
    let staticFullUrl;
    const isStatic = src.startsWith('/static/');
    const isExpandableAsset = src.startsWith('/assets/') && editorType === 'expandable';
    if ((isStatic || isExpandableAsset) && imageUrls.length > 0) {
      const assetSrc = src.substring(0, src.indexOf('"'));
      const assetName = assetSrc.replace(/\/assets\/.+[^/]\//g, '');
      const staticName = assetSrc.substring(8);
      imageUrls.forEach((url) => {
        if (isExpandableAsset && assetName === url.displayName) {
          staticFullUrl = `${lmsEndpointUrl}${url.staticFullUrl}`;
        } else if (staticName === url.displayName) {
          staticFullUrl = url.staticFullUrl;
          if (isExpandableAsset) {
            staticFullUrl = `${lmsEndpointUrl}${url.staticFullUrl}`;
          }
        }
      });
      if (staticFullUrl) {
        const currentSrc = src.substring(0, src.indexOf('"'));
        content = currentContent.replace(currentSrc, staticFullUrl);
        if (editorType === 'expandable') {
          updateContent(content);
        } else {
          editor.setContent(content);
        }
      }
    }
  });
};

export const getImageResizeHandler = ({ editor, imagesRef, setImage }) => () => {
  const {
    src, alt, width, height,
  } = editor.selection.getNode();

  imagesRef.current = module.updateImageDimensions({
    images: imagesRef.current, url: src, width, height,
  }).result;

  setImage({
    externalUrl: src,
    altText: alt,
    width,
    height,
  });
};

export const setupCustomBehavior = ({
  updateContent,
  openImgModal,
  openSourceCodeModal,
  editorType,
  imageUrls,
  images,
  setImage,
  lmsEndpointUrl,
}) => (editor) => {
  // image upload button
  editor.ui.registry.addButton(tinyMCE.buttons.imageUploadButton, {
    icon: 'image',
    tooltip: 'Add Image',
    onAction: openImgModal,
  });
  // editing an existing image
  editor.ui.registry.addButton(tinyMCE.buttons.editImageSettings, {
    icon: 'image',
    tooltip: 'Edit Image Settings',
    onAction: module.openModalWithSelectedImage({
      editor, images, setImage, openImgModal,
    }),
  });
  // overriding the code plugin's icon with 'HTML' text
  editor.ui.registry.addButton(tinyMCE.buttons.code, {
    text: 'HTML',
    tooltip: 'Source code',
    onAction: openSourceCodeModal,
  });
  // add a custom simple inline code block formatter.
  const setupCodeFormatting = (api) => {
    editor.formatter.formatChanged(
      'code',
      (active) => api.setActive(active),
    );
  };
  const toggleCodeFormatting = () => {
    editor.formatter.toggle('code');
    editor.undoManager.add();
    editor.focus();
  };
  editor.ui.registry.addToggleButton(tinyMCE.buttons.codeBlock, {
    icon: 'sourcecode',
    tooltip: 'Code Block',
    onAction: toggleCodeFormatting,
    onSetup: setupCodeFormatting,
  });
  // add a custom simple inline label formatter.
  const toggleLabelFormatting = () => {
    editor.execCommand('mceToggleFormat', false, 'label');
  };
  editor.ui.registry.addIcon('textToSpeech', tinyMCE.textToSpeechIcon);
  editor.ui.registry.addButton('customLabelButton', {
    icon: 'textToSpeech',
    text: 'Label',
    tooltip: 'Apply a "Question" label to specific text, recognized by screen readers. Recommended to improve accessibility.',
    onAction: toggleLabelFormatting,
  });
  if (editorType === 'expandable') {
    editor.on('init', () => {
      module.replaceStaticwithAsset({
        editor,
        imageUrls,
        editorType,
        lmsEndpointUrl,
        updateContent,
      });
    });
  }
  editor.on('ExecCommand', (e) => {
    if (editorType === 'text' && e.command === 'mceFocus') {
      module.replaceStaticwithAsset({ editor, imageUrls });
    }
    if (e.command === 'RemoveFormat') {
      editor.formatter.remove('blockquote');
      editor.formatter.remove('label');
    }
  });
  // after resizing an image in the editor, synchronize React state and ref
  editor.on('ObjectResized', getImageResizeHandler({ editor, imagesRef: images, setImage }));
};

// imagetools_cors_hosts needs a protocol-sanatized url
export const removeProtocolFromUrl = (url) => url.replace(/^https?:\/\//, '');

export const editorConfig = ({
  editorType,
  setEditorRef,
  editorContentHtml,
  images,
  lmsEndpointUrl,
  studioEndpointUrl,
  isLibrary,
  placeholder,
  initializeEditor,
  openImgModal,
  openSourceCodeModal,
  setSelection,
  updateContent,
  content,
  minHeight,
}) => {
  const {
    toolbar,
    config,
    plugins,
    imageToolbar,
    quickbarsInsertToolbar,
    quickbarsSelectionToolbar,
  } = pluginConfig({ isLibrary, placeholder, editorType });
  const isLocaleRtl = isRtl(getLocale());

  return {
    onInit: (evt, editor) => {
      setEditorRef(editor);
      if (editorType === 'text') {
        initializeEditor();
      }
    },
    initialValue: editorContentHtml || '',
    init: {
      ...config,
      skin: false,
      content_css: false,
      content_style: tinyMCEStyles + a11ycheckerCss,
      min_height: minHeight,
      contextmenu: 'link table',
      directionality: isLocaleRtl ? 'rtl' : 'ltr',
      document_base_url: lmsEndpointUrl,
      imagetools_cors_hosts: [removeProtocolFromUrl(lmsEndpointUrl), removeProtocolFromUrl(studioEndpointUrl)],
      imagetools_toolbar: imageToolbar,
      formats: { label: { inline: 'label' } },
      setup: module.setupCustomBehavior({
        editorType,
        updateContent,
        openImgModal,
        openSourceCodeModal,
        lmsEndpointUrl,
        setImage: setSelection,
        content,
        images,
        imageUrls: module.fetchImageUrls(images),
      }),
      quickbars_insert_toolbar: quickbarsInsertToolbar,
      quickbars_selection_toolbar: quickbarsSelectionToolbar,
      quickbars_image_toolbar: false,
      toolbar,
      plugins,
      valid_children: '+body[style]',
      valid_elements: '*[*]',
      entity_encoding: 'utf-8',
    },
  };
};

export const prepareEditorRef = () => {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const editorRef = useRef(null);
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const setEditorRef = useCallback((ref) => {
    editorRef.current = ref;
  }, []);
  const [refReady, setRefReady] = module.state.refReady(false);
  // eslint-disable-next-line react-hooks/rules-of-hooks
  useEffect(() => setRefReady(true), []);
  return { editorRef, refReady, setEditorRef };
};

export const imgModalToggle = () => {
  const [isImgOpen, setIsOpen] = module.state.isImageModalOpen(false);
  return {
    isImgOpen,
    openImgModal: () => setIsOpen(true),
    closeImgModal: () => setIsOpen(false),
  };
};

export const sourceCodeModalToggle = (editorRef) => {
  const [isSourceCodeOpen, setIsOpen] = module.state.isSourceCodeModalOpen(false);
  return {
    isSourceCodeOpen,
    openSourceCodeModal: () => setIsOpen(true),
    closeSourceCodeModal: () => {
      setIsOpen(false);
      editorRef.current.focus();
    },
  };
};

/**
 * const imageMatchRegex
 *
 * Image urls and ids used in the TinyMceEditor vary wildly, with different base urls,
 * different lengths and constituent parts, and replacement of some "/" with "@".
 * Common are the keys "asset-v1", "type", and "block", each holding a value after some separator.
 * This regex captures only the values for these keys using capture groups, which can be used for matching.
 */
export const imageMatchRegex = /asset-v1.(.*).type.(.*).block.(.*)/;

/**
 * function matchImageStringsByIdentifiers
 *
 * matches two strings by comparing their regex capture groups using the `imageMatchRegex`
 */
export const matchImageStringsByIdentifiers = (a, b) => {
  if (!a || !b || !(typeof a === 'string') || !(typeof b === 'string')) { return null; }
  const matchA = JSON.stringify(a.match(imageMatchRegex)?.slice?.(1));
  const matchB = JSON.stringify(b.match(imageMatchRegex)?.slice?.(1));
  return matchA && matchA === matchB;
};

export const stringToFragment = (htmlString) => document.createRange().createContextualFragment(htmlString);

export const getImageFromHtmlString = (htmlString, imageSrc) => {
  const images = stringToFragment(htmlString)?.querySelectorAll('img') || [];

  return Array.from(images).find((img) => matchImageStringsByIdentifiers(img.src || '', imageSrc));
};

export const detectImageMatchingError = ({ matchingImages, tinyMceHTML }) => {
  if (!matchingImages.length) { return true; }
  if (matchingImages.length > 1) { return true; }

  if (!matchImageStringsByIdentifiers(matchingImages[0].id, tinyMceHTML.src)) { return true; }
  if (!matchingImages[0].width || !matchingImages[0].height) { return true; }
  if (matchingImages[0].width !== tinyMceHTML.width) { return true; }
  if (matchingImages[0].height !== tinyMceHTML.height) { return true; }

  return false;
};

export const openModalWithSelectedImage = ({
  editor, images, setImage, openImgModal,
}) => () => {
  const tinyMceHTML = editor.selection.getNode();
  const { src: mceSrc } = tinyMceHTML;

  const matchingImages = images.current.filter(image => matchImageStringsByIdentifiers(image.id, mceSrc));

  const imageMatchingErrorDetected = detectImageMatchingError({ tinyMceHTML, matchingImages });

  const width = imageMatchingErrorDetected ? null : matchingImages[0]?.width;
  const height = imageMatchingErrorDetected ? null : matchingImages[0]?.height;

  setImage({
    externalUrl: tinyMceHTML.src,
    altText: tinyMceHTML.alt,
    width,
    height,
  });

  openImgModal();
};

export const filterAssets = ({ assets }) => {
  let images = [];
  const assetsList = Object.values(assets);
  if (assetsList.length > 0) {
    images = assetsList.filter(asset => asset?.contentType?.startsWith('image/'));
  }
  return images;
};

export const setAssetToStaticUrl = ({ editorValue, assets, lmsEndpointUrl }) => {
  /* For assets to remain usable across course instances, we convert their url to be course-agnostic.
   * For example, /assets/course/<asset hash>/filename gets converted to /static/filename. This is
   * important for rerunning courses and importing/exporting course as the /static/ part of the url
   * allows the asset to be mapped to the new course run.
  */

  // TODO: should probably move this to when the assets are being looped through in the off chance that
  // some of the text in the editor contains the lmsEndpointUrl
  const regExLmsEndpointUrl = RegExp(lmsEndpointUrl, 'g');
  let content = editorValue.replace(regExLmsEndpointUrl, '');

  const assetUrls = [];
  const assetsList = Object.values(assets);
  assetsList.forEach(asset => {
    assetUrls.push({ portableUrl: asset.portableUrl, displayName: asset.displayName });
  });
  const assetSrcs = typeof content === 'string' ? content.split(/(src="|src=&quot;|href="|href=&quot)/g) : [];
  assetSrcs.forEach(src => {
    if (src.startsWith('/asset') && assetUrls.length > 0) {
      const assetBlockName = src.substring(src.indexOf('@') + 1, src.search(/("|&quot;)/));
      const nameFromEditorSrc = assetBlockName.substring(assetBlockName.indexOf('@') + 1);
      const nameFromStudioSrc = assetBlockName.substring(assetBlockName.indexOf('/') + 1);
      let portableUrl;
      assetUrls.forEach((url) => {
        const displayName = url.displayName.replace(/\s/g, '_');
        if (displayName === nameFromEditorSrc || displayName === nameFromStudioSrc) {
          portableUrl = url.portableUrl;
        }
      });
      if (portableUrl) {
        const currentSrc = src.substring(0, src.search(/("|&quot;)/));
        const updatedContent = content.replace(currentSrc, portableUrl);
        content = updatedContent;
      }
    }
  });
  return content;
};

export const fetchImageUrls = (images) => {
  const imageUrls = [];
  images.current.forEach(image => {
    imageUrls.push({ staticFullUrl: image.staticFullUrl, displayName: image.displayName });
  });
  return imageUrls;
};

export const selectedImage = (val) => {
  const [selection, setSelection] = module.state.imageSelection(val);
  return {
    clearSelection: () => setSelection(null),
    selection,
    setSelection,
  };
};

/**
 * function updateImageDimensions
 *
 * Updates one images' dimensions in an array by identifying one image via a url string match
 * that includes asset-v1, type, and block. Returns a new array.
 *
 * @param {Object[]} images - [{ id, ...other }]
 * @param {string} url
 * @param {number} width
 * @param {number} height
 *
 * @returns {Object} { result, foundMatch }
 */
export const updateImageDimensions = ({
  images, url, width, height,
}) => {
  let foundMatch = false;

  const result = images.map((image) => {
    const imageIdentifier = image.id || image.url || image.src || image.externalUrl;
    const isMatch = matchImageStringsByIdentifiers(imageIdentifier, url);
    if (isMatch) {
      foundMatch = true;
      return { ...image, width, height };
    }
    return image;
  });

  return { result, foundMatch };
};
