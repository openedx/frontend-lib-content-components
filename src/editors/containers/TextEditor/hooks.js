import * as appHooks from '../../hooks';
import { setAssetToStaticUrl } from '../../sharedComponents/TinyMceWidget/hooks';

export const { nullMethod, navigateCallback, navigateTo } = appHooks;

export const getContent = ({ editorRef, isRawEditor }) => () => {
  const content = (isRawEditor && editorRef && editorRef.current
    ? editorRef.current.state.doc.toString()
    : editorRef.current?.getContent());
  return setAssetToStaticUrl({ editorValue: content });
};
