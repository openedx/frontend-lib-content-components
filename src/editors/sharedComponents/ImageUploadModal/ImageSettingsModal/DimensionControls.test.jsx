import React, { useEffect } from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import * as paragon from '@edx/paragon';
import * as icons from '@edx/paragon/icons';

import {
  fireEvent, render, screen, waitFor,
} from '@testing-library/react';
import { formatMessage } from '../../../../testUtils';
import { DimensionControls } from './DimensionControls';
import hooks from './hooks';

const WrappedDimensionControls = () => {
  const dimensions = hooks.dimensions('altText');

  useEffect(() => {
    dimensions.onImgLoad({ })({ target: { naturalWidth: 1517, naturalHeight: 803 } });
  }, []);

  return <DimensionControls {...dimensions} intl={{ formatMessage }} />;
};

const UnlockedDimensionControls = () => {
  const dimensions = hooks.dimensions('altText');

  useEffect(() => {
    dimensions.onImgLoad({ })({ target: { naturalWidth: 1517, naturalHeight: 803 } });
    dimensions.unlock();
  }, []);

  return <DimensionControls {...dimensions} intl={{ formatMessage }} />;
};

describe('DimensionControls', () => {
  describe('render', () => {
    const props = {
      lockAspectRatio: { width: 4, height: 5 },
      locked: { 'props.locked': 'lockedValue' },
      isLocked: true,
      value: { width: 20, height: 40 },
      // inject
      intl: { formatMessage },
    };
    beforeEach(() => {
      jest.spyOn(hooks, 'onInputChange').mockImplementation((handler) => ({ 'hooks.onInputChange': handler }));
      props.setWidth = jest.fn().mockName('props.setWidth');
      props.setHeight = jest.fn().mockName('props.setHeight');
      props.lock = jest.fn().mockName('props.lock');
      props.unlock = jest.fn().mockName('props.unlock');
      props.updateDimensions = jest.fn().mockName('props.updateDimensions');
    });
    afterEach(() => {
      jest.spyOn(hooks, 'onInputChange').mockRestore();
    });
    test('snapshot', () => {
      expect(shallow(<DimensionControls {...props} />).snapshot).toMatchSnapshot();
    });
    test('null value: empty snapshot', () => {
      const el = shallow(<DimensionControls {...props} value={null} />);
      expect(el.snapshot).toMatchSnapshot();
      expect(el.isEmptyRender()).toEqual(true);
    });
    test('unlocked dimensions', () => {
      const el = shallow(<DimensionControls {...props} isLocked={false} />);
      expect(el.snapshot).toMatchSnapshot();
    });
  });
  describe('component tests for dimensions', () => {
    beforeEach(() => {
      paragon.Form.Group = jest.fn().mockImplementation(({ children }) => (
        <div>{children}</div>
      ));
      paragon.Form.Label = jest.fn().mockImplementation(({ children }) => (
        <div>{children}</div>
      ));
      // eslint-disable-next-line no-import-assign
      paragon.Icon = jest.fn().mockImplementation(({ children }) => (
        <div>{children}</div>
      ));
      // eslint-disable-next-line no-import-assign
      paragon.IconButton = jest.fn().mockImplementation(({ children }) => (
        <div>{children}</div>
      ));
      paragon.Form.Control = jest.fn().mockImplementation(({ value, onChange, onBlur }) => (
        <input className="formControl" onChange={onChange} onBlur={onBlur} value={value} />
      ));
      // eslint-disable-next-line no-import-assign
      icons.Locked = jest.fn().mockImplementation(() => {});
      // eslint-disable-next-line no-import-assign
      icons.Unlocked = jest.fn().mockImplementation(() => {});
    });
    afterEach(() => {
      paragon.Form.Group.mockRestore();
      paragon.Form.Label.mockRestore();
      paragon.Form.Control.mockRestore();
      paragon.Icon.mockRestore();
      paragon.IconButton.mockRestore();
      icons.Locked.mockRestore();
      icons.Unlocked.mockRestore();
    });

    it('renders with initial dimensions', () => {
      const { container } = render(<WrappedDimensionControls />);
      const widthInput = container.querySelector('.formControl');
      expect(widthInput.value).toBe('1517');
    });

    it('resizes dimensions proportionally', async () => {
      const { container } = render(<WrappedDimensionControls />);
      const widthInput = container.querySelector('.formControl');
      expect(widthInput.value).toBe('1517');
      fireEvent.change(widthInput, { target: { value: 758 } });
      await waitFor(() => {
        expect(container.querySelectorAll('.formControl')[0].value).toBe('758');
      });
      fireEvent.blur(widthInput);
      await waitFor(() => {
        expect(container.querySelectorAll('.formControl')[0].value).toBe('758');
        expect(container.querySelectorAll('.formControl')[1].value).toBe('401');
      });
      screen.debug();
    });

    it('resizes only changed dimension when unlocked', async () => {
      const { container } = render(<UnlockedDimensionControls />);
      const widthInput = container.querySelector('.formControl');
      expect(widthInput.value).toBe('1517');
      fireEvent.change(widthInput, { target: { value: 758 } });
      await waitFor(() => {
        expect(container.querySelectorAll('.formControl')[0].value).toBe('758');
      });
      fireEvent.blur(widthInput);
      await waitFor(() => {
        expect(container.querySelectorAll('.formControl')[0].value).toBe('758');
        expect(container.querySelectorAll('.formControl')[1].value).toBe('803');
      });
      screen.debug();
    });
  });
});
