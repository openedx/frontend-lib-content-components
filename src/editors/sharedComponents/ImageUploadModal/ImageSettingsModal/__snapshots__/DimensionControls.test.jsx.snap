// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DimensionControls render null value: empty snapshot 1`] = `false`;

exports[`DimensionControls render snapshot 1`] = `
<Form.Group>
  <Form.Label
    as="h4"
  >
    <FormattedMessage
      defaultMessage="Image Dimensions"
      description="Label title for the image dimensions section."
      id="authoring.texteditor.imagesettingsmodal.imageDimensionsLabel"
    />
  </Form.Label>
  <div
    className="mt-4.5"
  >
    <Form.Control
      className="dimension-input"
      floatingLabel="Width"
      onBlur={[MockFunction props.updateDimensions]}
      onChange={
        Object {
          "hooks.onInputChange": [MockFunction props.setWidth],
        }
      }
      value={20}
    />
    <Form.Control
      className="dimension-input"
      floatingLabel="Height"
      onBlur={[MockFunction props.updateDimensions]}
      onChange={
        Object {
          "hooks.onInputChange": [MockFunction props.setHeight],
        }
      }
      value={40}
    />
    <IconButton
      alt="unlock dimensions"
      className="d-inline-block"
      iconAs="Icon"
      onClick={[MockFunction props.unlock]}
      src={[MockFunction icons.Locked]}
    />
  </div>
</Form.Group>
`;

exports[`DimensionControls render unlocked dimensions 1`] = `
<Form.Group>
  <Form.Label
    as="h4"
  >
    <FormattedMessage
      defaultMessage="Image Dimensions"
      description="Label title for the image dimensions section."
      id="authoring.texteditor.imagesettingsmodal.imageDimensionsLabel"
    />
  </Form.Label>
  <div
    className="mt-4.5"
  >
    <Form.Control
      className="dimension-input"
      floatingLabel="Width"
      onBlur={[MockFunction props.updateDimensions]}
      onChange={
        Object {
          "hooks.onInputChange": [MockFunction props.setWidth],
        }
      }
      value={20}
    />
    <Form.Control
      className="dimension-input"
      floatingLabel="Height"
      onBlur={[MockFunction props.updateDimensions]}
      onChange={
        Object {
          "hooks.onInputChange": [MockFunction props.setHeight],
        }
      }
      value={40}
    />
    <IconButton
      alt="lock dimensions"
      className="d-inline-block"
      iconAs="Icon"
      onClick={[MockFunction props.lock]}
      src={[MockFunction icons.Unlocked]}
    />
  </div>
</Form.Group>
`;
